File: hw2/python_implementation/recursive_ll_db.py
import os
import threading
import time
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

head = None
watched_file = None
watcher_thread = None
observer = None

class Node:
    def __init__(self, fields):
        self.fields = fields
        self.next = None

def load_csv_recursive(file_obj):
    line = file_obj.readline()
    if not line:
        return None
    fields = line.strip().split(",")
    node = Node(fields)
    node.next = load_csv_recursive(file_obj)
    return node

def print_list(head):
    current = head
    row_num = 1
    while current:
        print(f"Row {row_num}: {' | '.join(current.fields)}")
        current = current.next
        row_num += 1

def save_to_file(filename):
    global head
    if head is None:
        print("No data to save.")
        return

    try:
        with open(filename, 'w') as f:
            current = head
            while current:
                f.write(','.join(current.fields) + '\n')
                current = current.next
        print(f"Data saved to {filename}")
    except Exception as e:
        print(f"Failed to save file: {e}")

class FileChangeHandler(FileSystemEventHandler):
    def __init__(self, target_file):
        super().__init__()
        self.target_file = os.path.abspath(target_file)

    def on_modified(self, event):
        global head
        if os.path.abspath(event.src_path) == self.target_file:
            print("\n[INFO] Detected file change. Reloading...")
            try:
                with open(self.target_file, 'r') as f:
                    head = load_csv_recursive(f)
                print("[INFO] Reload complete.")
                print("> ", end="", flush=True)
            except Exception as e:
                print(f"[ERROR] Reload failed: {e}")

def start_file_watcher(filepath):
    global observer, watcher_thread
    dir_to_watch = os.path.dirname(os.path.abspath(filepath))
    handler = FileChangeHandler(filepath)

    observer = Observer()
    observer.schedule(handler, path=dir_to_watch, recursive=False)
    observer.start()

def stop_file_watcher():
    global observer
    if observer:
        observer.stop()
        observer.join()

def main():
    global head, watched_file

    print("In-Memory CSV Database")
    print("Commands: load, unload, print, save, exit")

    while True:
        try:
            command = input("> ").strip().lower()
        except (EOFError, KeyboardInterrupt):
            command = "exit"

        if command == "load":
            if head:
                print("A file is already loaded. Unload it first.")
                continue

            filename = input("Enter CSV filename: ").strip()
            if not os.path.exists(filename):
                print("File not found.")
                continue

            confirm = input(f"Confirm loading file '{filename}'? (yes/no): ").strip().lower()
            if confirm not in ("yes", "y"):
                print("Load cancelled.")
                continue

            try:
                with open(filename, 'r') as f:
                    head = load_csv_recursive(f)
                    watched_file = filename
                    print("File loaded into memory.")
                    start_file_watcher(filename)
            except Exception as e:
                print(f"Error loading file: {e}")

        elif command == "save":
            if not head:
                print("No data to save.")
                continue

            yn = input(f"Save to original file ({os.path.basename(watched_file)})? (Y/N): ").strip().lower()
            if yn in ("y", "yes"):
                save_to_file(watched_file)
            elif yn in ("n", "no"):
                new_file = input("Enter new filename to save as: ").strip()
                if new_file:
                    save_to_file(new_file)
                else:
                    print("Save cancelled: no filename given.")
            else:
                print("Invalid input. Save cancelled.")

        elif command == "unload":
            if not head:
                print("No file is currently loaded.")
                continue
            head = None
            watched_file = None
            stop_file_watcher()
            print("In-memory database unloaded.")

        elif command == "print":
            if not head:
                print("No data loaded.")
            else:
                print_list(head)

        elif command == "exit":
            print("Goodbye!")
            stop_file_watcher()
            break

        else:
            print("Unknown command. Try: load, unload, print, save, exit")

if __name__ == "__main__":
    main()

File: hw2/java_implementation/recursive_ll_db.java
import java.io.*;
import java.nio.file.*;
import java.util.Scanner;

public class recursive_ll_db {

    static volatile Node head = null;
    static File watchedFile = null;
    static Thread watcherThread = null;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("In-Memory CSV Database");
        System.out.println("Commands: load, unload, print, exit");

        while (true) {
            System.out.print("> ");
            String command = scanner.nextLine().trim().toLowerCase();

            switch (command) {
                case "load":
                    if (head != null) {
                        System.out.println("A file is already loaded. Unload it first.");
                        break;
                    }

                    System.out.print("Enter CSV filename: ");
                    String filename = scanner.nextLine().trim();
                    File file = new File(filename);

                    if (!file.exists() || file.isDirectory()) {
                        System.out.println("File not found.");
                        break;
                    }

                    System.out.print("Confirm loading file '" + filename + "'? (yes/no): ");
                    String confirm = scanner.nextLine().trim().toLowerCase();
                    if (!confirm.equals("yes") && !confirm.equals("y")) {
                        System.out.println("Load cancelled.");
                        break;
                    }

                    try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
                        head = loadCSV(reader);
                        watchedFile = file;
                        System.out.println("File loaded into memory.");
                        //startFileWatcher(file.toPath().getParent(), file.getName());

                        Path parentDir = file.toPath().getParent();
                        if (parentDir == null) {
                            // file is in current directory, use current working directory instead
                            parentDir = Paths.get(".");
                        }
                        startFileWatcher(parentDir.toAbsolutePath(), file.getName());

                    } catch (IOException e) {
                        System.out.println("Error loading file: " + e.getMessage());
                    }
                    break;

                case "save":
                    if (head == null) {
                        System.out.println("No data to save.");
                        break;
                    }

                    System.out.print("Save to original file (" + watchedFile.getName() + ")? (Y/N): ");
                    String yn = scanner.nextLine().trim().toLowerCase();

                    if (yn.equals("y") || yn.equals("yes")) {
                        saveToFile(watchedFile.getAbsolutePath());
                    } else if (yn.equals("n") || yn.equals("no")) {
                        System.out.print("Enter new filename to save as: ");
                        String newFile = scanner.nextLine().trim();
                        if (!newFile.isEmpty()) {
                            saveToFile(newFile);
                            // Optional: update watchedFile and restart watcher to new file
                            // For simplicity, not changing current watched file here
                        } else {
                            System.out.println("Save cancelled: no filename given.");
                        }
                    } else {
                        System.out.println("Invalid input. Save cancelled.");
                    }
                    break;

                case "unload":
                    if (head == null) {
                        System.out.println("No file is currently loaded.");
                        break;
                    }
                    head = null;
                    watchedFile = null;
                    if (watcherThread != null && watcherThread.isAlive()) {
                        watcherThread.interrupt(); // Stop the watcher
                    }
                    System.out.println("In-memory database unloaded.");
                    break;

                case "print":
                    if (head == null) {
                        System.out.println("No data loaded.");
                    } else {
                        printList(head);
                    }
                    break;

                case "exit":
                    System.out.println("Goodbye!");
                    if (watcherThread != null && watcherThread.isAlive()) {
                        watcherThread.interrupt();
                    }
                    return;

                default:
                    System.out.println("Unknown command. Try: load, unload, print, exit.");
            }
        }
    }

    // ================================
    // Node class to represent each row
    static class Node {
        String[] fields;
        Node next;

        public Node(String[] fields) {
            this.fields = fields;
            this.next = null;
        }
    }

    // ================================
    // Recursive CSV loader
    public static Node loadCSV(BufferedReader reader) throws IOException {
        String line = reader.readLine();
        if (line == null) return null; // Base case

        String[] fields = line.split(","); // Simple split
        Node node = new Node(fields);
        node.next = loadCSV(reader); // Recursive call
        return node;
    }

    // ================================
    // Print the linked list
    public static void printList(Node head) {
        Node current = head;
        int rowNum = 1;
        while (current != null) {
            System.out.print("Row " + rowNum + ": ");
            for (int i = 0; i < current.fields.length; i++) {
                System.out.print(current.fields[i]);
                if (i < current.fields.length - 1) System.out.print(" | ");
            }
            System.out.println();
            current = current.next;
            rowNum++;
        }
    }

    // ================================
    // File watcher
    public static void startFileWatcher(Path pathToWatch, String filename) {
        watcherThread = new Thread(() -> {
            try {
                WatchService watchService = FileSystems.getDefault().newWatchService();
                pathToWatch.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);

                while (!Thread.currentThread().isInterrupted()) {
                    WatchKey key = watchService.take(); // blocking
                    for (WatchEvent<?> event : key.pollEvents()) {
                        WatchEvent.Kind<?> kind = event.kind();
                        Path changed = (Path) event.context();

                        if (kind == StandardWatchEventKinds.ENTRY_MODIFY &&
                                changed.toString().equals(filename)) {
                            System.out.println("\n[INFO] Detected file change. Reloading...");

                            try (BufferedReader reader = new BufferedReader(new FileReader(watchedFile))) {
                                head = loadCSV(reader);
                                System.out.println("[INFO] Reload complete.");
                                System.out.print("\n> ");
                            } catch (IOException e) {
                                System.out.println("[ERROR] Reload failed: " + e.getMessage());
                            }
                        }
                    }
                    key.reset();
                }

            } catch (IOException | InterruptedException e) {
                if (!(e instanceof InterruptedException)) {
                    System.out.println("[ERROR] Watcher stopped: " + e.getMessage());
                }
            }
        });

        watcherThread.setDaemon(true);
        watcherThread.start();
    }

    // ================================
    // Save File
    public static void saveToFile(String filename) {
        if (head == null) {
            System.out.println("No data to save.");
            return;
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            Node current = head;
            while (current != null) {
                String line = String.join(",", current.fields);
                writer.write(line);
                writer.newLine();
                current = current.next;
            }
            System.out.println("Data saved to " + filename);
        } catch (IOException e) {
            System.out.println("Failed to save file: " + e.getMessage());
        }
    }
}
